src/utils/PluginDriver.ts

```diff
@@ -131,13 +132,15 @@ export class PluginDriver {
 		hookName: H,
 		parameters: Parameters<FunctionPluginHooks[H]>,
 		replaceContext?: ReplaceContext | null,
-		skipped?: ReadonlySet<Plugin> | null
+		skipped?: ReadonlySet<Plugin> | null,
+		callBeforeRunHook?: CallBeforeRunHook
 	): Promise<ReturnType<FunctionPluginHooks[H]> | null> {
```

Comment:
I wonder about the `callBeforeRunHook`. It is very powerful, but I currently do not see many other uses for it except to figure out the responsible plugin. Moreover, it adds another function call into every resolveId hook, which is one of the most-called hooks. I wonder if it would not be more performant to actually return the plugin from the hook?

Suggestion (which will also improve performance further by breaking out of the loop once we found a handler. This code obviously predates async-await): Add a new handler e.g.

```ts
async hookFirstAndGetPlugin<H extends AsyncPluginHooks & FirstPluginHooks>(
 		hookName: H,
 		parameters: Parameters<FunctionPluginHooks[H]>,
 		replaceContext?: ReplaceContext | null,
 		skipped?: ReadonlySet<Plugin> | null
 	): Promise<[result: ReturnType<FunctionPluginHooks[H]>, plugin: Plugin] | null> {
for (const plugin of this.getSortedPlugins(hookName)) {
			if (skipped && skipped.has(plugin)) continue;
			const result = await this.runHook(hookName, parameters, plugin, replaceContext);
      // Note the single "=" to match both null and undefined
			if (result != null) {
				return [result, plugin];
			}
		}
		return null;
	}
}
```

to avoid changing too many code places, we can define a new `hookFirst` that internally calls `hookFirstAndGetPlugin` and returns the first element if the result is not `null`. What do you think?

Reply:
Thanks for your review, i think your idea is a better resolution, i will do it.